File: ui/src/lib/api/costAttribution.ts
/**
 * Typed Cost Attribution client (P3).
 * Stable surface for /api/v1/cost-attribution/* until OpenAPI regen.
 */

import { client } from "@/client/client.gen";

export type CostGroupBy = "workflow" | "campaign" | "definition";

export type CostBucket = {
  key: string;
  label: string;
  group_type: "workflow" | "campaign" | "definition" | "unattributed";
  workflow_id: number | null;
  campaign_id: number | null;
  definition_id: number | null;
  run_count: number;
  runs_with_cost: number;
  runs_missing_cost: number;
  total_duration_seconds: number;
  total_cost_usd: number | null;
  total_charge_usd: number | null;
  total_dograh_tokens: number;
  avg_cost_usd: number | null;
  cost_coverage_pct: number;
};

export type CostAttributionSummary = {
  from_date: string;
  to_date: string;
  timezone: string;
  workflow_id: number | null;
  campaign_id: number | null;
  group_by: CostGroupBy;
  total_runs: number;
  runs_with_cost: number;
  runs_missing_cost: number;
  cost_coverage_pct: number;
  total_duration_seconds: number;
  total_cost_usd: number | null;
  total_charge_usd: number | null;
  total_dograh_tokens: number;
  buckets: CostBucket[];
  notes: string[];
};

export type CostAttributionQuery = {
  from_date: string;
  to_date: string;
  timezone?: string;
  workflow_id?: number | null;
  campaign_id?: number | null;
  group_by?: CostGroupBy;
};

function errorMessage(err: unknown, fallback: string): string {
  if (typeof err === "object" && err && "detail" in err) {
    return String((err as { detail: unknown }).detail);
  }
  if (err instanceof Error) return err.message;
  return fallback;
}

export async function fetchCostAttributionSummary(
  q: CostAttributionQuery,
): Promise<CostAttributionSummary> {
  const res = await client.get({
    url: "/api/v1/cost-attribution/summary",
    query: {
      from_date: q.from_date,
      to_date: q.to_date,
      timezone: q.timezone || "UTC",
      group_by: q.group_by || "workflow",
      ...(q.workflow_id != null ? { workflow_id: q.workflow_id } : {}),
      ...(q.campaign_id != null ? { campaign_id: q.campaign_id } : {}),
    },
  });
  if (res.error) {
    throw new Error(errorMessage(res.error, "Cost attribution summary failed"));
  }
  return res.data as CostAttributionSummary;
}

File: api/schemas/cost_attribution.py
"""Schemas for Cost Attribution dashboard (P3)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class CostBucket(BaseModel):
    """One aggregation dimension row (workflow / campaign / definition)."""

    key: str
    label: str
    group_type: Literal["workflow", "campaign", "definition", "unattributed"]
    workflow_id: int | None = None
    campaign_id: int | None = None
    definition_id: int | None = None
    run_count: int = 0
    runs_with_cost: int = 0
    runs_missing_cost: int = 0
    total_duration_seconds: float = 0.0
    total_cost_usd: float | None = None
    total_charge_usd: float | None = None
    total_dograh_tokens: float = 0.0
    avg_cost_usd: float | None = None
    cost_coverage_pct: float = 0.0


class CostAttributionSummary(BaseModel):
    from_date: str
    to_date: str
    timezone: str
    workflow_id: int | None = None
    campaign_id: int | None = None
    group_by: Literal["workflow", "campaign", "definition"] = "workflow"
    total_runs: int = 0
    runs_with_cost: int = 0
    runs_missing_cost: int = 0
    cost_coverage_pct: float = 0.0
    total_duration_seconds: float = 0.0
    total_cost_usd: float | None = None
    total_charge_usd: float | None = None
    total_dograh_tokens: float = 0.0
    buckets: list[CostBucket] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    total_matching_runs: int = 0
    sampled_runs: int = 0
    sample_limit: int = 0
    truncated: bool = False
    truncation_note: str | None = None
